File: python/packages/core/src/varar_core/registry.py
"""Registry for step definitions — port of var-core/src/registry.ts."""
from __future__ import annotations

from dataclasses import dataclass, field
from re import Pattern
from typing import Any, Callable, Mapping, Optional, Union

from cucumber_expressions.expression import CucumberExpression
from cucumber_expressions.parameter_type import ParameterType
from cucumber_expressions.parameter_type_registry import ParameterTypeRegistry

from varar_core.step_role import StepKind

# A step handler: receives the context state plus any matched arguments
StepHandler = Callable[..., Any]

# Accepted regexp forms for define_parameter_type
RegexpInput = Union[str, Pattern, list[Union[str, Pattern]]]

# A parameter type's display formatter: value -> the document's notation.
# Presentation only — never part of matching or comparison verdicts.
ParameterFormat = Callable[[Any], str]


@dataclass(frozen=True)
class StepRegistration:
    expression: str
    expression_source_file: str
    expression_source_line: int
    handler: StepHandler
    compiled: CucumberExpression
    kind: Optional[StepKind] = None


@dataclass(frozen=True)
class Registry:
    steps: tuple[StepRegistration, ...]
    parameter_types: ParameterTypeRegistry
    # Per parameter-type display formatters, keyed by type name. Kept beside
    # the cucumber-expressions registry because ParameterType can't carry one.
    formats: Mapping[str, ParameterFormat] = field(default_factory=dict)


# Markdown emphasis, as a built-in {emph} parameter type. Matches the uniform
# emphasis notations (bold-italic, bold, italic; ``*`` and ``_`` delimiters),
# ordered longest-delimiter-first so ``**x**`` isn't half-eaten by the ``*``
# branch. Each branch captures the inner text in its own group, so only the
# outermost delimiter pair is stripped (``**_x_**`` -> ``_x_``) and editors
# highlight the value, not the markers.
EMPH_REGEXP = r"\*\*\*([^*]+)\*\*\*|___([^_]+)___|\*\*([^*]+)\*\*|__([^_]+)__|\*([^*]+)\*|_([^_]+)_"


def _seed_builtins(registry: Registry) -> Registry:
    """Seed var's own built-in parameter types (beyond cucumber-expressions'
    int/float/string/word). Shared by every port so specs match identically.
    """
    return define_parameter_type(
        registry,
        name="emph",
        regexp=EMPH_REGEXP,
        # Exactly one alternation branch matches, so exactly one group is set.
        parse=lambda *groups: next((g for g in groups if g is not None), ""),
        # Emphasis is distinctive notation; don't auto-suggest it in snippets.
        use_for_snippets=False,
        # Mismatch display renders the value back in single-asterisk emphasis.
        format=lambda value: f"*{value}*",
    )


def create_registry() -> Registry:
    """Return a registry seeded with var's built-in parameter types."""
    empty = Registry(steps=(), parameter_types=ParameterTypeRegistry(), formats={})
    return _seed_builtins(empty)


def add_step(
    registry: Registry,
    *,
    expression: str,
    expression_source_file: str,
    expression_source_line: int,
    handler: StepHandler,
    kind: Optional[StepKind] = None,
) -> Registry:
    """Compile *expression* and append it to *registry*.

    Returns a new Registry; the original is unchanged (ParameterTypeRegistry is
    shared by reference — see note in TS source — but never mutated here).

    Raises ValueError on duplicate expressions, mirroring the TS error message.
    """
    duplicate = next((s for s in registry.steps if s.expression == expression), None)
    if duplicate is not None:
        raise ValueError(
            f'duplicate step definition for "{expression}" at '
            f"{duplicate.expression_source_file}:{duplicate.expression_source_line} and "
            f"{expression_source_file}:{expression_source_line}"
        )
    compiled = CucumberExpression(expression, registry.parameter_types)
    reg = StepRegistration(
        expression=expression,
        expression_source_file=expression_source_file,
        expression_source_line=expression_source_line,
        handler=handler,
        compiled=compiled,
        kind=kind,
    )
    return Registry(
        steps=(*registry.steps, reg),
        parameter_types=registry.parameter_types,
        formats=registry.formats,
    )


def define_parameter_type(
    registry: Registry,
    *,
    name: str,
    regexp: RegexpInput,
    parse: Optional[Callable[..., Any]] = None,
    use_for_snippets: bool = True,
    prefer_for_regexp_match: bool = False,
    format: Optional[ParameterFormat] = None,
) -> Registry:
    """Register a custom parameter type with the shared ParameterTypeRegistry.

    The registry's ParameterTypeRegistry is mutated in place (same semantics as
    the TS implementation), so previously compiled CucumberExpressions also gain
    the new type.

    Returns the same Registry object (the mutation is in-place).
    """
    # Normalise regexp to a list so ParameterType always receives a list
    if isinstance(regexp, (str, Pattern)):
        regexps: list[str | Pattern] = [regexp]
    else:
        regexps = list(regexp)

    # Default parse: identity — return the first (and usually only) group
    effective_parse: Callable[..., Any] = parse if parse is not None else (
        lambda *groups: groups[0]
    )

    pt: ParameterType = ParameterType(
        name,
        regexps,
        None,
        effective_parse,
        use_for_snippets,
        prefer_for_regexp_match,
    )
    registry.parameter_types.define_parameter_type(pt)
    if format is None:
        # Return the same object — the mutation is in the ParameterTypeRegistry
        return registry
    return Registry(
        steps=registry.steps,
        parameter_types=registry.parameter_types,
        formats={**registry.formats, name: format},
    )

File: typescript/packages/core/src/registry.ts
import {
  CucumberExpression,
  ParameterType,
  ParameterTypeRegistry,
} from '@cucumber/cucumber-expressions'
import type { StepKind } from './step-role.ts'

export type StepHandler = (
  state: unknown,
  ...args: ReadonlyArray<unknown>
) => unknown | Promise<unknown>

export type StepRegistration = {
  readonly expression: string
  readonly expressionSourceFile: string
  readonly expressionSourceLine: number
  readonly handler: StepHandler
  readonly compiled: CucumberExpression
  readonly kind?: StepKind
}

// A parameter type's display formatter: value → the document's notation.
// Presentation only — never part of matching or comparison verdicts.
export type ParameterFormat = (value: unknown) => string

export type Registry = {
  readonly steps: ReadonlyArray<StepRegistration>
  readonly parameterTypes: ParameterTypeRegistry
  // Per parameter-type display formatters, keyed by type name. Kept beside
  // the cucumber-expressions registry because ParameterType can't carry one.
  readonly formats: ReadonlyMap<string, ParameterFormat>
}

// Markdown emphasis, as a built-in {emph} parameter type. Matches the uniform
// emphasis notations (bold-italic, bold, italic; `*` and `_` delimiters),
// ordered longest-delimiter-first so `**x**` isn't half-eaten by the `*`
// branch. Each branch captures the inner text in its own group, so only the
// outermost delimiter pair is stripped (`**_x_**` → `_x_`) and editors
// highlight the value, not the markers.
export const EMPH_REGEXP =
  /\*\*\*([^*]+)\*\*\*|___([^_]+)___|\*\*([^*]+)\*\*|__([^_]+)__|\*([^*]+)\*|_([^_]+)_/

// Seed Varar's own built-in parameter types (beyond cucumber-expressions'
// int/float/string/word). Shared by every port so specs match identically.
function seedBuiltins(registry: Registry): Registry {
  return defineParameterType(registry, {
    name: 'emph',
    regexp: EMPH_REGEXP,
    // Exactly one alternation branch matches, so exactly one group is defined.
    parse: (...groups: ReadonlyArray<string | undefined>) =>
      groups.find((g) => g !== undefined) ?? '',
    // Emphasis is distinctive notation; don't auto-suggest it in snippets.
    useForSnippets: false,
    // Mismatch display renders the value back in single-asterisk emphasis.
    format: (value) => `*${String(value)}*`,
  })
}

export function createRegistry(): Registry {
  return seedBuiltins({
    steps: [],
    parameterTypes: new ParameterTypeRegistry(),
    formats: new Map(),
  })
}

export type StepInput = Omit<StepRegistration, 'compiled'>

export function addStep(registry: Registry, input: StepInput): Registry {
  const duplicate = registry.steps.find((s) => s.expression === input.expression)
  if (duplicate) {
    throw new Error(
      `duplicate step definition for "${input.expression}" at ${duplicate.expressionSourceFile}:${duplicate.expressionSourceLine} and ${input.expressionSourceFile}:${input.expressionSourceLine}`,
    )
  }
  const compiled = new CucumberExpression(input.expression, registry.parameterTypes)
  const next: StepRegistration = { ...input, compiled }
  return { ...registry, steps: [...registry.steps, next] }
}

export type ParameterTypeInput<T = unknown> = {
  readonly name: string
  readonly regexp: RegExp | string | ReadonlyArray<RegExp | string>
  // Identity by default — turns the matched substring into the handler's
  // argument. Override to coerce (e.g. number, Date, domain object).
  readonly parse?: (...groups: string[]) => T
  // Whether this type can be auto-suggested when generating a snippet from
  // arbitrary text. Defaults to true.
  readonly useForSnippets?: boolean
  // Take priority over built-in {string}/{int}/etc. when generating a regex
  // for an existing string match. Rarely needed.
  readonly preferForRegexpMatch?: boolean
  // Inverse of `parse`: render a value in the document's notation.
  // Used only to display the actual side of a parameter mismatch.
  readonly format?: (value: T) => string
}

export function defineParameterType<T = unknown>(
  registry: Registry,
  input: ParameterTypeInput<T>,
): Registry {
  // ParameterTypeRegistry.defineParameterType mutates in place; the registry
  // object reference is intentionally shared across all step compilations.
  const pt = new ParameterType<T>(
    input.name,
    input.regexp as RegExp | string | RegExp[] | string[],
    null,
    input.parse ?? ((raw: string) => raw as unknown as T),
    input.useForSnippets ?? true,
    input.preferForRegexpMatch ?? false,
    false,
  )
  registry.parameterTypes.defineParameterType(pt)
  if (!input.format) return registry
  const formats = new Map(registry.formats)
  formats.set(input.name, input.format as ParameterFormat)
  return { ...registry, formats }
}
